src/inmanta/util.py

```diff
@@ -207,3 +208,93 @@ def custom_json_encoder(o: object) -> Union[Dict, str, List]:
+        Add a future to the ioloop to be handled, but do not require the result.
+    """
+    def handle_result(f: Task) -> None:
+        try:
+            f.result()
+        except Exception as e:
+            LOGGER.exception("An exception occurred while handling a future: %s", str(e))
+
+    task = ensure_future(future)
+    task.add_done_callback(handle_result)
+    return task
+
+
+async def retry_limited(fun: Callable[[], bool], timeout: float, interval: float = 0.1) -> None:
+    start = time.time()
+    while time.time() - start < timeout and not fun():
+        await sleep(interval)
+
+
+class TaskHandler(object):
+    """
+        This class provides a method to add a background task based on a coroutine. When the coroutine ends, any exceptions
+        are reported. If stop is invoked, all background tasks are cancelled.
+    """
+    def __init__(self) -> None:
+        super().__init__()
+        self._background_tasks: Set[Task] = set()
+        self._await_tasks: Set[Task] = set()
+
+    def add_future(self, future: Union[Future, Coroutine]) -> Task:
+        """ Add a future to the event loop and report any exceptions if they occur
+        """
+        def handle_result(f: Task) -> None:
+            try:
+                f.result()
+            except Exception as e:
+                LOGGER.exception("An exception occurred while handling a future: %s", str(e))
+
+        task = ensure_future(future)
+        task.add_done_callback(handle_result)
+        return task
+
+    def add_background_task(self, future: Union[Future, Coroutine], cancel_on_stop=True) -> None:
+        """ Add a background task to the event loop. When stop is called, the task is cancelled.
+
+        :param future: The future or coroutine to run as background task.
+        :param cancel_on_stop: Cancel the task when stop is called. If false, the coroutine is awaited.
+        """
+        task = ensure_future(future)
+
+        def handle_result(task: Task) -> None:
+            try:
+                task.result()
+            except CancelledError:
+                LOGGER.warning("Task %s was cancelled.", task)
+
+            except Exception as e:
+                LOGGER.exception("An exception occurred while handling a future: %s", str(e))
+            finally:
+                if task in self._background_tasks:
+                    self._background_tasks.remove(task)
+
+        task.add_done_callback(handle_result)
+        self._background_tasks.add(task)
+
+        if not cancel_on_stop:
+            self._await_tasks.add(task)
+
+        return task
+
+    async def stop(self) -> None:
+        """ Stop all background tasks by requestinng a cancel
+        """
+        await gather(*self._await_tasks)
+        self._background_tasks.difference_update(self._await_tasks)
+
+        try:
+            while True:
+                task = self._background_tasks.pop()
+                task.cancel()
+                try:
+                    await task
```

Comment:
1. why await them one by one, not cancel all available, then wait for all of them in parallel?
2. perhaps log a warning if new tasks are added during stop?

Reply:
1) Not sure what happens in gather. Will find out.
2) Log a warning and except because we should not add a task to the event loop and otherwise we might return none